### PR TITLE
Performance/838 multi threading

### PR DESCRIPTION
**Pull request by Nova-2119**
### Benefits

Speed improvements to opening graphs and how quickly the graph resets after opening a view.

### Possible Drawbacks

There may still be a multithreading bug hanging around in there. Unable to confirm as I could not get it to occur again.
People will be unable to go and get a coffee while their larger graphs load.

### Applicable Issues
#838

### CODE REVIEW

File: CoreOpenGLDisplay/src/au/gov/asd/tac/constellation/visual/opengl/renderer/batcher/NodeLabelBatcher.java [hunk 1]
```diff
@@ -145,99 +124,114 @@ public void createShader(GL3 gl) throws IOException {
...
-        final FloatArray bottomLabelFloats = new FloatArray();
-        final IntArray bottomLabelInts = new IntArray();
+    public GLRenderableUpdateTask createBatch(final VisualAccess access) throws InterruptedException {
+        final NodeGlyphStream topGlyphStream = new NodeGlyphStream();
+        final NodeGlyphStream bottomGlyphStream = new NodeGlyphStream();
+        
+        Thread topLabelThread = new FillTopLabels(access, topGlyphStream);
+        topLabelThread.start();
+        
+        Thread bottomLabelThread = new FillBottomLabels(access, bottomGlyphStream);
```
Comment (antares1470) [inline CoreOpenGLDisplay/src/au/gov/asd/tac/constellation/visual/opengl/renderer/batcher/NodeLabelBatcher.java:134]: This and the thread above can be final

File: CoreOpenGLDisplay/src/au/gov/asd/tac/constellation/visual/opengl/renderer/batcher/NodeLabelBatcher.java [hunk 2]
```diff
@@ -145,99 +124,114 @@ public void createShader(GL3 gl) throws IOException {
...
     }
 
-    public GLRenderableUpdateTask updateTopLabels(final VisualAccess access) {
+    public GLRenderableUpdateTask updateTopLabels(final VisualAccess access) throws InterruptedException {
         // We build the whole batch again - can't update labels in place at this stage.
-        final FloatArray topLabelFloats = new FloatArray();
-        final IntArray topLabelInts = new IntArray();
-        fillTopLabels(access, topLabelFloats, topLabelInts);
+        NodeGlyphStream glyphStream = new NodeGlyphStream();
+        Thread topLabelThread = new FillTopLabels(access, glyphStream);
```
Comment (antares1470) [inline CoreOpenGLDisplay/src/au/gov/asd/tac/constellation/visual/opengl/renderer/batcher/NodeLabelBatcher.java:155]: both final

File: CoreOpenGLDisplay/src/au/gov/asd/tac/constellation/visual/opengl/renderer/batcher/NodeLabelBatcher.java [hunk 3]
```diff
@@ -145,99 +124,114 @@ public void createShader(GL3 gl) throws IOException {
...
     }
 
-    public GLRenderableUpdateTask updateBottomLabels(final VisualAccess access) {
+    public GLRenderableUpdateTask updateBottomLabels(final VisualAccess access) throws InterruptedException {
         // We build the whole batch again - can't update labels in place at this stage.
-        final FloatArray bottomLabelFloats = new FloatArray();
-        final IntArray bottomLabelInts = new IntArray();
-        fillBottomLabels(access, bottomLabelFloats, bottomLabelInts);
+        NodeGlyphStream glyphStream = new NodeGlyphStream();
+        Thread bottomLabelThread = new FillBottomLabels(access, glyphStream);
```
Comment (antares1470) [inline CoreOpenGLDisplay/src/au/gov/asd/tac/constellation/visual/opengl/renderer/batcher/NodeLabelBatcher.java:172]: both final

File: CoreOpenGLDisplay/src/au/gov/asd/tac/constellation/visual/opengl/renderer/batcher/NodeLabelBatcher.java [hunk 4]
```diff
@@ -145,99 +124,114 @@ public void createShader(GL3 gl) throws IOException {
...
+            bottomBatch.buffer(gl, labelIntsTarget, IntBuffer.wrap(glyphStream.getCurrentInts().rawArray()));
             bottomBatch.finalise(gl);
         };
     }
     
-    private void fillTopLabels(final VisualAccess access, final FloatArray topLabelFloats, final IntArray topLabelInts) {
-        currentFloats = topLabelFloats;
-        currentInts = topLabelInts;
+    private void fillTopLabels(final VisualAccess access, NodeGlyphStream glyphStream) throws InterruptedException {
+        ExecutorService pool = Executors.newFixedThreadPool(NUM_CORES);
```
Comment (antares1470) [inline CoreOpenGLDisplay/src/au/gov/asd/tac/constellation/visual/opengl/renderer/batcher/NodeLabelBatcher.java:187]: final

File: CoreOpenGLDisplay/src/au/gov/asd/tac/constellation/visual/opengl/renderer/batcher/NodeLabelBatcher.java [hunk 5]
```diff
@@ -145,99 +124,114 @@ public void createShader(GL3 gl) throws IOException {
...
     }
     
-    private void fillTopLabels(final VisualAccess access, final FloatArray topLabelFloats, final IntArray topLabelInts) {
-        currentFloats = topLabelFloats;
-        currentInts = topLabelInts;
+    private void fillTopLabels(final VisualAccess access, NodeGlyphStream glyphStream) throws InterruptedException {
+        ExecutorService pool = Executors.newFixedThreadPool(NUM_CORES);  
         for (int pos = 0; pos < access.getVertexCount(); pos++) {
-            bufferTopLabel(pos, access);
+            Thread thread = new BufferTopLabel(pos, access, glyphStream);
```
Comment (antares1470) [inline CoreOpenGLDisplay/src/au/gov/asd/tac/constellation/visual/opengl/renderer/batcher/NodeLabelBatcher.java:189]: final

File: CoreOpenGLDisplay/src/au/gov/asd/tac/constellation/visual/opengl/renderer/batcher/NodeLabelBatcher.java [hunk 6]
```diff
@@ -145,99 +124,114 @@ public void createShader(GL3 gl) throws IOException {
...
 
-    private void fillBottomLabels(final VisualAccess access, final FloatArray bottomLabelFloats, final IntArray bottomLabelInts) {
-        currentFloats = bottomLabelFloats;
-        currentInts = bottomLabelInts;
+    private void fillBottomLabels(final VisualAccess access, NodeGlyphStream glyphStream) throws InterruptedException {
+        ExecutorService pool = Executors.newFixedThreadPool(NUM_CORES);  
+        
         for (int pos = 0; pos < access.getVertexCount(); pos++) {
-            bufferBottomLabel(pos, access);
+            Thread thread = new BufferBottomLabel(pos, access, glyphStream);
```
Comment (antares1470) [inline CoreOpenGLDisplay/src/au/gov/asd/tac/constellation/visual/opengl/renderer/batcher/NodeLabelBatcher.java:203]: this and `pool` can be final

File: CoreOpenGLDisplay/src/au/gov/asd/tac/constellation/visual/opengl/renderer/batcher/NodeLabelBatcher.java [hunk 7]
```diff
@@ -340,4 +334,75 @@ public void drawBatch(final GL3 gl, final Camera camera, final Matrix44f mvMatri
             }
         }
     }
-}
+
+    
+    class FillTopLabels extends Thread {
+        final private VisualAccess access;
+        final private NodeGlyphStream glyphStream;
```
Comment (antares1470) [inline CoreOpenGLDisplay/src/au/gov/asd/tac/constellation/visual/opengl/renderer/batcher/NodeLabelBatcher.java:341]: swap the modifiers around on both of these to be `private final`

File: CoreOpenGLDisplay/src/au/gov/asd/tac/constellation/visual/opengl/renderer/batcher/NodeLabelBatcher.java [hunk 8]
```diff
@@ -340,4 +334,75 @@ public void drawBatch(final GL3 gl, final Camera camera, final Matrix44f mvMatri
...
+                fillTopLabels(access, glyphStream);
+            } catch (InterruptedException ex) {
+                Exceptions.printStackTrace(ex);
+            }
+        }
+    }   
+    
+    class FillBottomLabels extends Thread {
+            final private VisualAccess access;
+            final private NodeGlyphStream glyphStream;
```
Comment (antares1470) [inline CoreOpenGLDisplay/src/au/gov/asd/tac/constellation/visual/opengl/renderer/batcher/NodeLabelBatcher.java:359]: same here

File: CoreOpenGLDisplay/src/au/gov/asd/tac/constellation/visual/opengl/renderer/batcher/NodeLabelBatcher.java [hunk 9]
```diff
@@ -340,4 +334,75 @@ public void drawBatch(final GL3 gl, final Camera camera, final Matrix44f mvMatri
...
+                } catch (InterruptedException ex) {
+                    Exceptions.printStackTrace(ex);
+                }
+            }
+    }
+    
+    class BufferBottomLabel extends Thread {
+        final private int pos;
+        final private VisualAccess access;
+        final private NodeGlyphStream glyphStream;
```
Comment (antares1470) [inline CoreOpenGLDisplay/src/au/gov/asd/tac/constellation/visual/opengl/renderer/batcher/NodeLabelBatcher.java:378]: same here

File: CoreOpenGLDisplay/src/au/gov/asd/tac/constellation/visual/opengl/renderer/batcher/NodeLabelBatcher.java [hunk 10]
```diff
@@ -340,4 +334,75 @@ public void drawBatch(final GL3 gl, final Camera camera, final Matrix44f mvMatri
...
+        public void run() {
+            bufferBottomLabel(pos, access, glyphStream);
+        }
+        
+    }
+
+    class BufferTopLabel extends Thread {
+        final private int pos;
+        final private VisualAccess access;
+        final private NodeGlyphStream glyphStream;
```
Comment (antares1470) [inline CoreOpenGLDisplay/src/au/gov/asd/tac/constellation/visual/opengl/renderer/batcher/NodeLabelBatcher.java:395]: same here

File: CoreOpenGLDisplay/src/au/gov/asd/tac/constellation/visual/opengl/utilities/glyphs/ConnectionGlyphStream.java [hunk 11]
```diff
@@ -0,0 +1,78 @@
...
+    public ConnectionGlyphStream() {
+        this.currentFloats = new FloatArray();
+        this.currentInts = new IntArray();
+    }
+    
+    
+    @Override
+    public void addGlyph(int glyphPosition, float x, float y, final GlyphStreamContext streamContext) {
+        if (streamContext instanceof ConnectionGlyphStreamContext) {
+            ConnectionGlyphStreamContext context = (ConnectionGlyphStreamContext) streamContext;
```
Comment (antares1470) [inline CoreOpenGLDisplay/src/au/gov/asd/tac/constellation/visual/opengl/utilities/glyphs/ConnectionGlyphStream.java:44]: final

File: CoreOpenGLDisplay/src/au/gov/asd/tac/constellation/visual/opengl/utilities/glyphs/ConnectionGlyphStream.java [hunk 12]
```diff
@@ -0,0 +1,78 @@
...
+                currentFloats.add(currentWidth, x, y, context.visibility);
+                currentInts.add(context.currentLowNodeId, context.currentHighNodeId, (context.currentOffset << 16) + (context.totalScale << 2) + context.labelNumber, (glyphPosition << 8) + context.currentStagger * 256 / (Math.min(context.currentLinkLabelCount, ConnectionLabelBatcher.MAX_STAGGERS) + 1));
+            }
+        }
+    }
+
+    @Override
+    public void newLine(float width, final GlyphStreamContext streamContext) {
+        if (streamContext instanceof ConnectionGlyphStreamContext) {
+            ConnectionGlyphStreamContext context = (ConnectionGlyphStreamContext) streamContext;
```
Comment (antares1470) [inline CoreOpenGLDisplay/src/au/gov/asd/tac/constellation/visual/opengl/utilities/glyphs/ConnectionGlyphStream.java:55]: final

File: CoreOpenGLDisplay/src/au/gov/asd/tac/constellation/visual/opengl/utilities/glyphs/NodeGlyphStream.java [hunk 13]
```diff
@@ -0,0 +1,82 @@
...
+    public NodeGlyphStream() {
+        this.currentFloats = new FloatArray();
+        this.currentInts = new IntArray();
+    }
+    
+    
+    @Override
+    public void addGlyph(final int glyphPosition, final float x, final float y, final GlyphStreamContext streamContext) {
+        if (streamContext instanceof NodeGlyphStreamContext){
+            NodeGlyphStreamContext context = (NodeGlyphStreamContext) streamContext;
```
Comment (antares1470) [inline CoreOpenGLDisplay/src/au/gov/asd/tac/constellation/visual/opengl/utilities/glyphs/NodeGlyphStream.java:42]: final

File: CoreOpenGLDisplay/src/au/gov/asd/tac/constellation/visual/opengl/utilities/glyphs/NodeGlyphStream.java [hunk 14]
```diff
@@ -0,0 +1,82 @@
...
+        }
+        else {
+            throw new IllegalArgumentException("Provided context lacks Node information, please use a NodeGlyphStreamContext");
+        }
+    }
+
+    @Override
+    public void newLine(float width, final GlyphStreamContext streamContext) {
+        if (streamContext instanceof NodeGlyphStreamContext) {
+            NodeGlyphStreamContext context = (NodeGlyphStreamContext) streamContext;
```
Comment (antares1470) [inline CoreOpenGLDisplay/src/au/gov/asd/tac/constellation/visual/opengl/utilities/glyphs/NodeGlyphStream.java:56]: final

**Review by antares1470 — CHANGES_REQUESTED** [review]

**Review by arcturus2 — APPROVED** [review]

**Review by aldebaran30701 — CHANGES_REQUESTED** [review]
I seem to have found a bug - the steps below are reproducers.
I followed the steps below quickly.
open new graph -> generate structured graph builder (1000) backbone size
open new graph -> generate sphere graph (10000)
open new graph -> generate sphere graph (1000)
While it was all trying to load, I switched between graphs and while it happened to try and load the sphere, I was on the structured graph. It loaded those elements onto that graph instead of the sphere graph. 
See the below gif for the behavior

render glitch

EDIT: I managed to get this reproduced in the current master. I'll make a ticket for it.
I'm happy for the functionality of this to be approved, I haven't taken a look at the code yet but will be providing my suggestions there also if needed.

**Review by aldebaran30701 — APPROVED** [review]
Didn't find any usage problems, waaay faster!
I didn't find any extra code smells after @antares1470  review. 
Great work ~

**Comment by arcturus2** [comment]
@antares1470 I've addressed your comments in 43355e7 and some code warnings in 1aea839

**Review by antares1470 — APPROVED** [review]
I could probably get more picky with cleaning up more of the code smells but I think we can merge this in now

**Pull request merged**